flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageProducerClientTest.java

```diff
@@ -0,0 +1,187 @@
+
+/** Tests for {@link TieredStorageProducerClient}. */
+@ExtendWith(ParameterizedTestExtension.class)
+public class TieredStorageProducerClientTest {
+
+    private static final int NUM_TOTAL_BUFFERS = 1000;
+
+    private static final int NETWORK_BUFFER_SIZE = 1024;
+
+    @Parameter public boolean isBroadcast;
+
+    private NetworkBufferPool globalPool;
+
+    @Parameters(name = "isBroadcast={0}")
+    public static Collection<Boolean> parameters() {
+        return Arrays.asList(false, true);
+    }
+
+    @BeforeEach
+    void before() {
+        globalPool = new NetworkBufferPool(NUM_TOTAL_BUFFERS, NETWORK_BUFFER_SIZE);
+    }
+
+    @AfterEach
+    void after() {
+        globalPool.destroy();
+    }
+
+    @TestTemplate
+    void testWriteRecordsToEmptyStorageTiers() {
+        int numSubpartitions = 10;
+        int bufferSize = 1024;
+        Random random = new Random();
+
+        TieredStorageProducerClient tieredStorageProducerClient =
+                createTieredStorageProducerClient(numSubpartitions, Collections.emptyList());
+        assertThatThrownBy(
+                        () ->
+                                tieredStorageProducerClient.write(
+                                        generateRandomData(bufferSize, random),
+                                        new TieredStorageSubpartitionId(0),
+                                        Buffer.DataType.DATA_BUFFER,
+                                        isBroadcast))
+                .isInstanceOf(RuntimeException.class)
+                .hasMessageContaining("Failed to choose a storage tier");
+    }
+
+    @TestTemplate
+    void testWriteRecords() throws IOException {
+        int numSubpartitions = 10;
+        int bufferSize = 1024;
+        int maxNumToWriteRecordsPerSubpartition = 1000;
+        Random random = new Random();
+
+        TestingTierProducerAgent tierProducerAgent = new TestingTierProducerAgent();
+        tierProducerAgent.setTryStartNewSegmentReturnValueSupplier(() -> true);
+        tierProducerAgent.setTryWriteReturnValueSupplier(() -> new Boolean[] {false, true});
+
+        TieredStorageProducerClient tieredStorageProducerClient =
+                createTieredStorageProducerClient(
+                        numSubpartitions, Collections.singletonList(tierProducerAgent));
+
+        int numWriteRecords = 0;
+        for (int j = 0; j < numSubpartitions; j++) {
+            for (int i = 0; i < random.nextInt(maxNumToWriteRecordsPerSubpartition); i++) {
+                tieredStorageProducerClient.write(
+                        generateRandomData(bufferSize, random),
+                        new TieredStorageSubpartitionId(j),
+                        Buffer.DataType.DATA_BUFFER,
+                        isBroadcast);
+                numWriteRecords++;
+            }
+        }
+
+        int numExpectedBuffers = isBroadcast ? numWriteRecords * numSubpartitions : numWriteRecords;
```

Comment:
I wonder if repeating write multiple times here is actually equivalent to `tryWrite (false) -> tryStartNewSegment -> tryWrite (true)`? 🤔

Reply:
Fixed. But in order to mock the scenario of writing failed, then choosing a new tier, then writing success, I created two testing tiers.